### PR TITLE
[io sites] adds versioned error view

### PR DESCRIPTION
**Pull request by BRKalow**
## "Ready for Review" checklist

- [ ] The Vercel preview link has been added to this PR's description
- [ ] The Asana task has been added to this PR's description
- [ ] This PR's link has been added to the "📐 Pull Request" field on the Asana task
- [ ] This Vercel preview link has been added to the "🔍 Preview" field on the Asana task
- [ ] The description template has been filled in below

---

## Relevant links

- [Preview link](url) 🔎
- [Asana task](url) 🎟️

## What

- Adds customized error views for versioned docs pages, as well as integrates the new `react-error-view` component

Screen Shot 2022-02-09 at 1 51 56 PM

## Testing

- [ ] Navigate to https://dev-portal-git-brkfeat-versioned-error-view-hashicorp.vercel.app
- [ ] Select waypoint in the bottom left
- [ ] Navigate to https://dev-portal-git-brkfeat-versioned-error-view-hashicorp.vercel.app/docs/v0.5.x/kubernetes/external-build
- [ ] Ensure the waypoint io layout is rendered
- [ ] Ensure the versioned error page is showing up (heading should say `This page does not exist for version v0.5.x.`)
- [ ] Click the "most recent version" link
- [ ] Observe navigated to `/docs/kubernetes/external-build`
- [ ] Click the "Go back to Documentation" link
- [ ] Observe navigated to `/docs`

### CODE REVIEW

File: src/pages/_app.tsx [hunk 1]
```diff
@@ -18,7 +18,7 @@ export default function App({ Component, pageProps }) {
   const Layout = Component.layout ?? BaseLayout
 
   return (
-    <ErrorBoundary FallbackComponent={Error}>
```
Comment (BRKalow) [inline src/pages/_app.tsx:21]: `Error` was the actual, native `Error` constructor, not a component 😟

File: src/pages/_error.jsx [hunk 2]
```diff
@@ -54,10 +53,21 @@ export async function getServerSideProps({ req, res, err }) {
   // Determine which layout to use, may be dev-portal's base layout,
   // or may be a proxied product layout, depending on the URL host
   const urlObj = new URL(req.url, `http://${req.headers.host}`)
-  const proxiedProductSlug = getProxiedProductSlug(urlObj.hostname)
+  // In preview environments, we can force the app into a certain .io mode with the io_preview cookie
+  const ioPreviewProduct =
+    process.env.HASHI_ENV === 'preview' ? req.cookies['io_preview'] : null
+
+  const proxiedProductSlug =
+    ioPreviewProduct ?? getProxiedProductSlug(urlObj.hostname, ioPreviewProduct)
```
Comment (BRKalow) [inline src/pages/_error.jsx]: Ensures that the correct product layout renders in deploy previews based on the `io_preview` cookie

File: src/views/_proxied-dot-io/versioned-error/index.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      </h1>
+      <p>
+        Please select either the{' '}
+        <Link href={pathWithoutVersion}>
+          <a>most recent version</a>
+        </Link>{' '}
+        or a valid version that includes the page you are looking for.
+      </p>
+      <p>
+        <Link href={`/${basePath}`}>
```
Comment (dstaley) [inline src/views/_proxied-dot-io/versioned-error/index.tsx:30]: I'm seeing

```html
<a href="/_proxied-dot-io">← Go back to Documentation</a>
```

Is this just a byproduct of the preview environment?

**Review by dstaley — APPROVED** [review]
LGTM, assuming the `<a href` thing is just a result of the preview environment.

File: src/views/_proxied-dot-io/versioned-error/index.tsx (hunk 3 above)
Comment (BRKalow) [reply src/views/_proxied-dot-io/versioned-error/index.tsx:30]: I noticed this too, debugged it for quite awhile today 😞  When clicking on the links it takes you to the correct place. And inspecting the components in the react devtools shows the right values within the `Link` component, but for some reason the actual href attribute isn't updated.

I'm hoping it'll be correct when deployed, but if not I'll circle back.

**Pull request merged**